src/Algebra/Bundles/Free/Magma.agda

```diff
@@ -0,0 +1,262 @@
+  sym s (r₁ ∙ r₂) = sym s r₁ ∙ sym s r₂
+
+  trans : Transitive R → Transitive _≈R_
+  trans t (var r)   (var s)   = var (t r s)
+  trans t (r₁ ∙ r₂) (s₁ ∙ s₂) = trans t r₁ s₁ ∙ trans t r₂ s₂
+
+  preservesEquivalence : IsEquivalence R → IsEquivalence _≈R_
+  preservesEquivalence isEq = record { refl = refl r ; sym = sym s ; trans = trans t }
+    where open IsEquivalence isEq renaming (refl to r; sym to s; trans to t)
+
+------------------------------------------------------------------------
+-- Free algebra on a Setoid
+
+module FreeMagmaOn {c ℓ} (S : Setoid c ℓ) where
+
+  open Setoid S renaming (Carrier to A; isEquivalence to isEq)
+
+  open PreFreeTheory _≈_ public
+
+  open PreservesEquivalence _≈_
+
+  open Strs _≈R_
+
+  isMagma : IsMagma  _∙_
+  isMagma = record { isEquivalence = preservesEquivalence isEq ; ∙-cong = _∙_ }
+
+  freeMagma : Magma c (c ⊔ ℓ)
+  freeMagma = record { Carrier = PreFreeMagma A; _≈_ = _≈R_ ; _∙_ = _∙_ ; isMagma = isMagma }
+
+------------------------------------------------------------------------
+-- Eval, as the unique fold ⟦_⟧_ over PreFreeMagma A
+
+module Eval {a ℓa m ℓm} (𝓐 : Setoid a ℓa) (𝓜 : Magma m ℓm) where
+
+  open Setoid 𝓐 renaming (Carrier to A)
+
+  open Magma 𝓜 renaming (Carrier to M; _∙_ to _∙ᴹ_)
+
+  ⟦_⟧_ : PreFreeMagma A → (A → M) → M
+  ⟦ var a ⟧ η = η a
+  ⟦ s ∙ t ⟧ η = ⟦ s ⟧ η ∙ᴹ ⟦ t ⟧ η
+
+------------------------------------------------------------------------
+-- Any Magma *is* an algebra for the PreFreeMagma Functor
+
+module Alg {m ℓm} (𝓜 : Magma m ℓm) where
+
+  open Magma 𝓜 renaming (setoid to setoidᴹ; Carrier to M)
+
+  open Eval setoidᴹ 𝓜
+
+  algᴹ : PreFreeMagma M → M
+  algᴹ t = ⟦ t ⟧ id
+
+------------------------------------------------------------------------
+-- Properties of ⟦_⟧_
+
+module Properties {a ℓa m ℓm} (𝓐 : Setoid a ℓa) (𝓜 : Magma m ℓm) where
+
+  open Setoid 𝓐 renaming (Carrier to A; _≈_ to _≈ᴬ_)
+
+  open Magma 𝓜
+    renaming (Carrier to M; _≈_ to _≈ᴹ_; _∙_ to _∙ᴹ_
+             ; setoid to setoidᴹ; rawMagma to rawMagmaᴹ; refl to reflᴹ
+             ; isMagma to isMagmaᴹ)
+
+  open Eval 𝓐 𝓜
+
+  open Alg 𝓜
+
+  open FreeMagmaOn 𝓐
+
+  open Magma freeMagma renaming (rawMagma to rawMagmaᴬ; Carrier to FA)
+
+  module _ {η : A → M} (hom-η : IsRelHomomorphism _≈ᴬ_ _≈ᴹ_ η) where
+
+    ⟦_⟧ᴹ : FA → M
+    ⟦_⟧ᴹ = ⟦_⟧ η
+
+    open Strs _≈ᴹ_
+    open IsMagma isMagmaᴹ renaming (∙-cong to congᴹ)
+    open IsRelHomomorphism hom-η renaming (cong to cong-η)
+
+    cong : ∀ {s t} → s ≈ t → ⟦ s ⟧ᴹ ≈ᴹ ⟦ t ⟧ᴹ
+    cong (var r) = cong-η r
+    cong (s ∙ t) = congᴹ (cong s) (cong t)
+
+    isRelHomomorphism : IsRelHomomorphism _≈_ _≈ᴹ_ ⟦_⟧ᴹ
+    isRelHomomorphism = record { cong = cong }
+
+    isMagmaHomomorphism : IsMagmaHomomorphism rawMagmaᴬ rawMagmaᴹ ⟦_⟧ᴹ
+    isMagmaHomomorphism = record { isRelHomomorphism = isRelHomomorphism
+                                 ; homo = λ _ _ → reflᴹ
+                                 }
+
+    unfold-⟦_⟧ᴹ : ∀ t → ⟦ t ⟧ᴹ ≈ᴹ algᴹ (map η t)
```

Comment:
Nice! Though I wonder if this is also known as 'zig' in the full proof of adjointness?

Reply:
Yes, perhaps. I'm never quite sure how people think about/talk about these special cases, but I guess I had `foldMap` at the back if my mind, but wanted a more neutral/descriptive name.

Reply:
IIRC, `algᴹ` is the (left) mate across the adjunction of the identity on the underlying `Setoid`. The `varSetoidHomomoprhism` is the `η` of the adjunction (I've overloaded `η`, perhaps unfortunately, because I want to use it for 'injection of generators' as well in the defintion of `Eval`), and there'll be (yet) more work to do to show that their compositions do the right thing...